certbot-apache/certbot_apache/dualparser.py

```diff
@@ -93,3 +93,140 @@ def set_parameters(self, parameters):
+        return new_comment
+
+    def _create_matching_list(self, primary_list, secondary_list):  # pragma: no cover
+        """ Matches the list of primary_list to a list of secondary_list and
+        returns a list of tuples. This is used to create results for find_
+        methods.
+
+        This helper function exists, because we cannot ensure that the list of
+        search results returned by primary.find_* and secondary.find_* are ordered
+        in a same way. The function pairs the same search results from both
+        implementations to a list of tuples.
+        """
+
+        matched = list()
+        for p in primary_list:
+            match = None
+            for s in secondary_list:
+                try:
+                    assertions.assertEqual(p, s)
+                    match = s
+                except AssertionError:
+                    continue
+            if match:
+                matched.append((p, match))
+            else:
+                raise AssertionError("Could not find a matching node.")
+        return matched
+
+    def find_blocks(self, name, exclude=True):  # pragma: no cover
+        """
+        Performs a search for BlockNodes using both implementations and does simple
+        checks for results. This is built upon the assumption that unimplemented
+        find_* methods return a list with a single assertion passing object.
+        After the assertion, it creates a list of newly created DualBlockNode
+        instances that encapsulate the pairs of returned BlockNode objects.
+        """
+        pass
+
+    def find_directives(self, name, exclude=True):  # pragma: no cover
+        """
+        Performs a search for DirectiveNodes using both implementations and
+        checks the results. This is built upon the assumption that unimplemented
+        find_* methods return a list with a single assertion passing object.
+        After the assertion, it creates a list of newly created DualDirectiveNode
+        instances that encapsulate the pairs of returned DirectiveNode objects.
+        """
+        pass
+
+    def find_comments(self, comment, exact=False):  # pragma: no cover
+        """
+        Performs a search for CommentNodes using both implementations and
+        checks the results. This is built upon the assumption that unimplemented
+        find_* methods return a list with a single assertion passing object.
+        After the assertion, it creates a list of newly created DualCommentNode
+        instances that encapsulate the pairs of returned CommentNode objects.
+        """
+        pass
+
+    def delete_child(self, child):  # pragma: no cover
+        """Deletes a child from the ParserNode implementations. The actual
+        ParserNode implementations are used here directly in order to be able
+        to match a child to the list of children."""
+
+        self.primary.delete_child(child.primary)
```

Comment:
Can this fail in one or the other of the implementations? Actually, this question probably applies to the `add_*` methods as well. Do we need to be raising and checking exceptions here?

Reply:
The interface doesn't currently define exceptions, so this will depend on the implementations themelves, and doing `except Exception` would in my opinion be overkill here and could hide some actual underlying issues. The execptions will definitely need to be handled here too when implementing them to the actual interface though.

Reply:
ok, I'm fine with leaving it out for now and then adding it as necessary later